File: packages/langchain-wrappers/src/marktoflow_langchain/copilot_llm.py
"""
GitHub Copilot CLI LangChain Wrapper

Uses GitHub Copilot CLI to provide LLM capabilities without separate API keys.
Requires GitHub Copilot subscription and `copilot` CLI installed.
"""

import subprocess
import logging
from typing import Any, List, Optional, Dict
from langchain_core.callbacks import CallbackManagerForLLMRun
from langchain_core.language_models.llms import LLM
from langchain_core.outputs import Generation, LLMResult

logger = logging.getLogger(__name__)


class GitHubCopilotLLM(LLM):
    """
    LangChain LLM wrapper for GitHub Copilot CLI.

    Uses your GitHub Copilot subscription - no separate API key needed.

    Example:
        ```python
        from marktoflow_langchain import GitHubCopilotLLM
        from browser_use import Agent

        llm = GitHubCopilotLLM(model="gpt-4.1")
        agent = Agent(
            task="Navigate to example.com and extract the title",
            llm=llm
        )
        result = agent.run()
        ```

    Requirements:
        - GitHub Copilot subscription
        - `copilot` CLI installed and authenticated
        - Install: npm install -g @github/copilot-cli
        - Authenticate: copilot auth
    """

    model: str = "gpt-4.1"
    """Model to use (gpt-4.1, gpt-4o, etc.)"""

    cli_path: str = "copilot"
    """Path to copilot CLI executable"""

    timeout: int = 120
    """Timeout for CLI calls in seconds"""

    verbose_output: bool = False
    """Enable verbose logging"""

    @property
    def _llm_type(self) -> str:
        """Return identifier for this LLM type."""
        return "github-copilot-cli"

    def _call(
        self,
        prompt: str,
        stop: Optional[List[str]] = None,
        run_manager: Optional[CallbackManagerForLLMRun] = None,
        **kwargs: Any,
    ) -> str:
        """
        Call GitHub Copilot CLI with the given prompt.

        Args:
            prompt: The prompt to send to Copilot
            stop: Stop sequences (not used by CLI)
            run_manager: Callback manager for streaming
            **kwargs: Additional arguments

        Returns:
            Response from Copilot CLI

        Raises:
            RuntimeError: If CLI call fails
        """
        if self.verbose_output:
            logger.info(f"Calling GitHub Copilot CLI with prompt: {prompt[:100]}...")

        # Build CLI command
        cmd = [self.cli_path, "-p", prompt]

        if self.model:
            cmd.extend(["--model", self.model])

        try:
            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=self.timeout,
                check=False
            )

            if result.returncode != 0:
                error_msg = f"GitHub Copilot CLI failed with code {result.returncode}"
                if result.stderr:
                    error_msg += f"\nSTDERR: {result.stderr}"
                if result.stdout:
                    error_msg += f"\nSTDOUT: {result.stdout}"
                raise RuntimeError(error_msg)

            response = result.stdout.strip()

            if self.verbose_output:
                logger.info(f"GitHub Copilot response: {response[:100]}...")

            return response

        except subprocess.TimeoutExpired as e:
            raise RuntimeError(
                f"GitHub Copilot CLI timed out after {self.timeout}s"
            ) from e
        except FileNotFoundError as e:
            raise RuntimeError(
                f"GitHub Copilot CLI not found at '{self.cli_path}'. "
                "Install it with: npm install -g @github/copilot-cli"
            ) from e
        except Exception as e:
            raise RuntimeError(f"GitHub Copilot CLI error: {e}") from e

    def _generate(
        self,
        prompts: List[str],
        stop: Optional[List[str]] = None,
        run_manager: Optional[CallbackManagerForLLMRun] = None,
        **kwargs: Any,
    ) -> LLMResult:
        """
        Generate responses for multiple prompts.

        Args:
            prompts: List of prompts
            stop: Stop sequences
            run_manager: Callback manager
            **kwargs: Additional arguments

        Returns:
            LLMResult with generations
        """
        generations = []
        for prompt in prompts:
            text = self._call(prompt, stop=stop, run_manager=run_manager, **kwargs)
            generations.append([Generation(text=text)])

        return LLMResult(generations=generations)

    @property
    def _identifying_params(self) -> Dict[str, Any]:
        """Return identifying parameters."""
        return {
            "model": self.model,
            "cli_path": self.cli_path,
            "timeout": self.timeout,
        }

    def check_installation(self) -> bool:
        """
        Check if GitHub Copilot CLI is installed and accessible.

        Returns:
            True if CLI is accessible, False otherwise
        """
        try:
            result = subprocess.run(
                [self.cli_path, "--version"],
                capture_output=True,
                text=True,
                timeout=5
            )
            return result.returncode == 0
        except (subprocess.TimeoutExpired, FileNotFoundError):
            return False

    def check_auth(self) -> bool:
        """
        Check if GitHub Copilot CLI is authenticated.

        Returns:
            True if authenticated, False otherwise
        """
        try:
            # Try a simple test prompt
            result = subprocess.run(
                [self.cli_path, "-p", "test"],
                capture_output=True,
                text=True,
                timeout=10
            )
            # If it returns without auth errors, we're authenticated
            return result.returncode == 0
        except (subprocess.TimeoutExpired, FileNotFoundError):
            return False

File: packages/langchain-adapters/src/copilot-llm.ts
/**
 * LangChain LLM wrapper for GitHub Copilot CLI
 *
 * Allows using GitHub Copilot subscription with LangChain.js
 * without requiring separate OpenAI API keys.
 *
 * @example
 * ```typescript
 * import { GitHubCopilotLLM } from '@marktoflow/langchain-adapters';
 *
 * const llm = new GitHubCopilotLLM({
 *   model: 'gpt-4.1',
 *   timeout: 120000,
 * });
 *
 * const response = await llm.invoke('What is the capital of France?');
 * console.log(response);
 * ```
 */

import { BaseLLM, type BaseLLMParams } from '@langchain/core/language_models/llms';
import { CallbackManagerForLLMRun } from '@langchain/core/callbacks/manager';
import type { LLMResult } from '@langchain/core/outputs';
import { spawn } from 'child_process';

export interface GitHubCopilotLLMParams extends BaseLLMParams {
  /**
   * Model to use (e.g., 'gpt-4.1', 'gpt-4o', 'gpt-4o-mini')
   * @default 'gpt-4.1'
   */
  model?: string;

  /**
   * Path to the copilot CLI executable
   * @default 'copilot'
   */
  cliPath?: string;

  /**
   * Timeout in milliseconds
   * @default 120000 (2 minutes)
   */
  timeout?: number;

  /**
   * Enable verbose logging
   * @default false
   */
  verbose?: boolean;
}

/**
 * LangChain LLM wrapper for GitHub Copilot CLI
 *
 * Uses your GitHub Copilot subscription instead of requiring OpenAI API keys.
 *
 * Requirements:
 * - GitHub Copilot subscription (~$10/month)
 * - Copilot CLI installed: npm install -g @github/copilot-cli
 * - Authenticated: copilot auth
 */
export class GitHubCopilotLLM extends BaseLLM {
  model: string;
  cliPath: string;
  timeout: number;
  verboseOutput: boolean;

  constructor(params: GitHubCopilotLLMParams = {}) {
    super(params);
    this.model = params.model ?? 'gpt-4.1';
    this.cliPath = params.cliPath ?? 'copilot';
    this.timeout = params.timeout ?? 120000;
    this.verboseOutput = params.verbose ?? false;
  }

  _llmType(): string {
    return 'github-copilot-cli';
  }

  /**
   * Call the GitHub Copilot CLI
   */
  async _call(
    prompt: string,
    _options?: this['ParsedCallOptions'],
    _runManager?: CallbackManagerForLLMRun,
  ): Promise<string> {
    const args = ['-p', prompt];

    if (this.model) {
      args.push('--model', this.model);
    }

    if (this.verboseOutput) {
      console.log(`[GitHubCopilotLLM] Calling CLI: ${this.cliPath} ${args.join(' ')}`);
    }

    return new Promise((resolve, reject) => {
      const timeoutId = setTimeout(() => {
        process.kill();
        reject(new Error(`GitHub Copilot CLI timed out after ${this.timeout}ms`));
      }, this.timeout);

      const process = spawn(this.cliPath, args, {
        stdio: ['pipe', 'pipe', 'pipe'],
      });

      let stdout = '';
      let stderr = '';

      process.stdout?.on('data', (data) => {
        stdout += data.toString();
      });

      process.stderr?.on('data', (data) => {
        stderr += data.toString();
      });

      process.on('error', (error) => {
        clearTimeout(timeoutId);
        reject(new Error(`Failed to spawn GitHub Copilot CLI: ${error.message}`));
      });

      process.on('close', (code) => {
        clearTimeout(timeoutId);

        if (code !== 0) {
          reject(new Error(
            `GitHub Copilot CLI failed with code ${code}:\n${stderr || 'No error output'}`
          ));
          return;
        }

        const output = stdout.trim();

        if (this.verboseOutput) {
          console.log(`[GitHubCopilotLLM] Response: ${output.substring(0, 100)}...`);
        }

        resolve(output);
      });
    });
  }

  /**
   * Generate method for LangChain compatibility
   */
  async _generate(
    prompts: string[],
    _options?: this['ParsedCallOptions'],
    _runManager?: CallbackManagerForLLMRun,
  ): Promise<LLMResult> {
    const generations = await Promise.all(
      prompts.map(async (prompt) => {
        const text = await this._call(prompt, _options, _runManager);
        return [{ text }];
      })
    );

    return { generations };
  }

  /**
   * Check if Copilot CLI is installed
   */
  async checkInstallation(): Promise<boolean> {
    try {
      await new Promise<void>((resolve, reject) => {
        const process = spawn(this.cliPath, ['--version'], {
          stdio: ['pipe', 'pipe', 'pipe'],
        });

        process.on('error', reject);
        process.on('close', (code) => {
          if (code === 0) {
            resolve();
          } else {
            reject(new Error(`CLI exited with code ${code}`));
          }
        });
      });
      return true;
    } catch (error) {
      return false;
    }
  }

  /**
   * Check if Copilot CLI is authenticated
   */
  async checkAuth(): Promise<boolean> {
    try {
      const response = await this._call('test');
      return response.length > 0;
    } catch (error) {
      return false;
    }
  }
}
